File: android/src/main/java/com/amitrana/rnbdk/RnBdkModule.kt
package com.ltbl.rnbdk

import android.util.Log
import com.facebook.react.bridge.*
import org.bitcoindevkit.*

val TAG = "RN-BDK"

object Progress : BdkProgress {
    override fun update(progress: Float, message: String?) {
        Log.i("Progress", "Sync wallet $progress $message")
    }
}

class RnBdkModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {
    override fun getName() = "RnBdkModule"
    override fun getConstants(): MutableMap<String, Any> {
        return hashMapOf("count" to 1)
    }

    // BDK module's real properties and methods
    private val externalDescriptor =
        "wpkh([c258d2e4/84h/1h/0h]tpubDDYkZojQFQjht8Tm4jsS3iuEmKjTiEGjG6KnuFNKKJb5A6ZUCUZKdvLdSDWofKi4ToRCwb9poe1XdqfUnP4jaJjCB2Zwv11ZLgSbnZSNecE/0/*)"
    private val internalDescriptor =
        "wpkh([c258d2e4/84h/1h/0h]tpubDDYkZojQFQjht8Tm4jsS3iuEmKjTiEGjG6KnuFNKKJb5A6ZUCUZKdvLdSDWofKi4ToRCwb9poe1XdqfUnP4jaJjCB2Zwv11ZLgSbnZSNecE/1/*)"
    private val databaseConfig = DatabaseConfig.Memory("")
    private val blockchainConfig =
        BlockchainConfig.Electrum(
            ElectrumConfig("ssl://electrum.blockstream.info:60002", null, 5u, null, 10u)
        )
    private lateinit var wallet: Wallet
    private var nodeNetwork = Network.TESTNET

    // Init wallet
    init {
        this.wallet = Wallet(
            externalDescriptor,
            internalDescriptor,
            nodeNetwork,
            databaseConfig,
            blockchainConfig
        )
        this.wallet.sync(Progress, null)
    }

    fun _seed(
        recover: Boolean = false,
        mnemonic: String = "",
        password: String? = null
    ): ExtendedKeyInfo {
        return if (!recover) generateExtendedKey(
            nodeNetwork,
            WordCount.WORDS12,
            password
        ) else restoreExtendedKey(nodeNetwork, mnemonic, password)
    }

    private fun createRestoreWallet(keys: ExtendedKeyInfo): Wallet {
        try {
            val descriptor: String = createDescriptor(keys)
            val changeDescriptor: String = createChangeDescriptor(keys)
            val newWallet = Wallet(
                descriptor,
                changeDescriptor,
                nodeNetwork,
                databaseConfig,
                blockchainConfig
            )
            newWallet.sync(Progress, null)
            return newWallet
        } catch (error: Error) {
            throw error
        }
    }

    private fun createDescriptor(keys: ExtendedKeyInfo): String {
        return ("wpkh(" + keys.xprv + "/84'/1'/0'/0/*)")
    }

    private fun createChangeDescriptor(keys: ExtendedKeyInfo): String {
        return ("wpkh(" + keys.xprv + "/84'/1'/0'/1/*)")
    }

    @ReactMethod
    fun genSeed(password: String?, promise: Promise) {
        try {
            val seed = _seed(false)
            promise.resolve(seed.mnemonic)
        } catch (error: Error) {
            return promise.reject("Gen Seed Error", error.localizedMessage, error)
        }
    }

    @ReactMethod
    fun createWallet(mnemonic: String = "", password: String?, promise: Promise) {
        try {
            val keys: ExtendedKeyInfo =
                _seed(if (mnemonic != "") true else false, mnemonic, password)
            val newWallet = createRestoreWallet(keys)
            promise.resolve("Address: ${newWallet.getNewAddress()}, Mnemonic: ${keys.mnemonic}")
        } catch (error: Error) {
            return promise.reject("Create Wallet Error", error.localizedMessage, error)
        }
    }

    @ReactMethod
    fun restoreWallet(mnemonic: String, password: String? = null, promise: Promise) {
        try {
            val keys: ExtendedKeyInfo = _seed(true, mnemonic, password)
            val newWallet = createRestoreWallet(keys)
            promise.resolve("Balance: ${newWallet.getBalance()} Address: ${newWallet.getNewAddress()}")
        } catch (error: Error) {
            return promise.reject("Restore Wallet Error", error.localizedMessage, error)
        }
    }

    @ReactMethod
    fun getNewAddress(promise: Promise) {
        promise.resolve(this.wallet.getNewAddress())
    }

    @ReactMethod
    fun getBalance(promise: Promise) {
        try {
            val balance = this.wallet.getBalance().toString()
            promise.resolve(balance)
        } catch (error: Error) {
            return promise.reject("Get Balance Error", error.localizedMessage, error)
        }
    }

    @ReactMethod
    fun broadcastTx(recepient: String, amount: Integer, promise: Promise) {
        try {
            val longAmt: Long = amount.toLong();
            val keys: ExtendedKeyInfo = _seed(
                true,
                "cushion merry upper hat mind tip fly ritual scheme civil disease since",
                null
            )
            val newWal: Wallet = createRestoreWallet(keys)
            Log.i("Custom Logs=======", "")
            Log.i(TAG, newWal.getNewAddress());
            Log.i(TAG, newWal.getBalance().toString())
            Log.i("=======Custom Logs", "")

            val psbt =
                PartiallySignedBitcoinTransaction(newWal, recepient, longAmt.toULong(), null)
            newWal.sign(psbt)

            val transaction: Transaction = newWal.broadcast(psbt)

            val details = when (transaction) {
                is Transaction.Confirmed -> transaction.details
                is Transaction.Unconfirmed -> transaction.details
            }

            val txidString = details.txid

            Log.i("TxID", "Transaction was broadcast! txid: $txidString")
            Log.i("TxID", "Transaction was broadcast!!!")
        } catch (error: Error) {
            return promise.reject("Transaction Error", error.localizedMessage, error)
        }
    }
}

File: ios/RnBdkModule.swift
//
//  RnBdkModule.swift
//  RnBdkModule
//

//
import Foundation

class Progress : BdkProgress {
    func update(progress: Float, message: String?) {
        print("progress", progress, message as Any)
    }
}

let TAG = "RN-BDK"

@objc(RnBdkModule)
class RnBdkModule: NSObject {
    
    let descriptor = "wpkh([c258d2e4/84h/1h/0h]tpubDDYkZojQFQjht8Tm4jsS3iuEmKjTiEGjG6KnuFNKKJb5A6ZUCUZKdvLdSDWofKi4ToRCwb9poe1XdqfUnP4jaJjCB2Zwv11ZLgSbnZSNecE/0/*)"
    let changeDescriptor = "wpkh([c258d2e4/84h/1h/0h]tpubDDYkZojQFQjht8Tm4jsS3iuEmKjTiEGjG6KnuFNKKJb5A6ZUCUZKdvLdSDWofKi4ToRCwb9poe1XdqfUnP4jaJjCB2Zwv11ZLgSbnZSNecE/1/*)"
    let databaseConfig = DatabaseConfig.memory(junk: "")
    let blockchainConfig = BlockchainConfig.electrum(
        config: ElectrumConfig(url: "ssl://electrum.blockstream.info:60002", socks5: nil, retry: 5, timeout: nil, stopGap: 10))
    var wallet: Wallet;
    let nodeNetwork = Network.testnet
    
    @objc static func requiresMainQueueSetup() -> Bool {
        return false
    }

    override init() {
        self.wallet = try! Wallet.init(descriptor: descriptor, changeDescriptor: changeDescriptor, network: nodeNetwork, databaseConfig: databaseConfig, blockchainConfig: blockchainConfig)
        try! self.wallet.sync(progressUpdate: Progress(), maxAddressParam: nil)
    }

    func _seed(
        recover: Bool = false,
        mnemonic: String?,  
        password: String? = nil
    ) -> ExtendedKeyInfo {
        return !recover ? try! generateExtendedKey(network: nodeNetwork, wordCount:WordCount.words12, password: password)
        : try! restoreExtendedKey(network: nodeNetwork, mnemonic: mnemonic ?? "", password: password)
    }


    private func createRestoreWallet(keys: ExtendedKeyInfo) throws -> Wallet {
        do{
            let descriptor: String = createDescriptor(keys: keys)
            let changeDescriptor: String = createChangeDescriptor(keys: keys)
            let newWallet: Wallet = try Wallet(descriptor: descriptor, changeDescriptor: changeDescriptor, network: nodeNetwork, databaseConfig: databaseConfig, blockchainConfig: blockchainConfig)
            try newWallet.sync(progressUpdate: Progress(), maxAddressParam: nil)
            return newWallet
        }
        catch {
            return error as! Wallet
        }
    }

    private func createDescriptor(keys: ExtendedKeyInfo)-> String {
        return ("wpkh(" + keys.xprv + "/84'/1'/0'/0/*)")
    }

    private func createChangeDescriptor(keys: ExtendedKeyInfo)-> String {
        return ("wpkh(" + keys.xprv + "/84'/1'/0'/1/*)")
    }

    @objc
    func genSeed(_ password: String? = nil, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let seed = _seed(recover: false, mnemonic: "")
        resolve(seed.mnemonic)
    }

    @objc
    func createWallet(_ mnemonic: String? = "", password: String? = nil, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            let keys: ExtendedKeyInfo = _seed(recover: mnemonic != "" ? true : false, mnemonic: mnemonic, password: password)
            let newWallet = try createRestoreWallet(keys: keys)
            resolve("Address: \(newWallet.getNewAddress()), \n Mnemonic: \(keys.mnemonic)")
        }
        catch {
            return reject("Create Wallet Error", error.localizedDescription, error)
        }
    }

    @objc
    func restoreWallet(_ mnemonic: String, password: String? = nil, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            let keys: ExtendedKeyInfo = _seed(recover: true, mnemonic: mnemonic, password: password)
            let newWallet = try createRestoreWallet(keys: keys)
            resolve("Balance: \(try newWallet.getBalance()) Address: \(newWallet.getNewAddress())")
        }
        catch {
            return reject("Restore Wallet Error", error.localizedDescription, error)
        }
    }

    @objc
    func getNewAddress(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        resolve(self.wallet.getNewAddress())
    }

    @objc
    func getBalance(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            let balance = try self.wallet.getBalance()
            resolve(balance)
        }
        catch {
            return reject("Get Balance Error", error.localizedDescription, error)
        }
    }

    @objc
    func broadcastTx(_ recipient: String, amount: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock){
        do {
            let keys: ExtendedKeyInfo = _seed(recover: true, mnemonic: "cushion merry upper hat mind tip fly ritual scheme civil disease since", password: nil)
            let newWal: Wallet = try createRestoreWallet(keys: keys)
            let psbt: PartiallySignedBitcoinTransaction = try PartiallySignedBitcoinTransaction(wallet: newWal, recipient: recipient, amount: UInt64(truncating: amount), feeRate: nil)
            try newWal.sign(psbt: psbt)
            let transaction = try newWal.broadcast(psbt: psbt)
            print("Broadcast success", transaction)
            resolve(recipient)
        }
        catch let error {
            return reject("Transaction Error", error.localizedDescription, error)
        }
    }

}


// "Address: tb1qc0pnkezmhcpgt70lu5djpl8pyypmfl7tyt5jcs, \n Mnemonic: cushion merry upper hat mind tip fly ritual scheme civil disease since"
